src/ovs/aca_arp_responder.cpp

```diff
@@ -0,0 +1,405 @@
+
+  return EXIT_SUCCESS;
+}
+
+
+void ACA_ARP_Responder::_validate_mac_address(const char *mac_string)
+{
+  unsigned char mac[6];
+
+  if (!mac_string) {
+    throw std::invalid_argument("Input mac_string is null");
+  }
+
+  if (sscanf(mac_string, "%hhx:%hhx:%hhx:%hhx:%hhx:%hhx", &mac[0], &mac[1],
+             &mac[2], &mac[3], &mac[4], &mac[5]) == 6) {
+    return;
+  }
+
+  if (sscanf(mac_string, "%hhx-%hhx-%hhx-%hhx-%hhx-%hhx", &mac[0], &mac[1],
+             &mac[2], &mac[3], &mac[4], &mac[5]) == 6) {
+    return;
+  }
+
+  // nothing matched
+  ACA_LOG_ERROR("Invalid mac address: %s\n", mac_string);
+
+  throw std::invalid_argument("Input mac_string is not in the expect format");
+}
+
+void ACA_ARP_Responder::_validate_ipv4_address(const char *ip_address)
+{
+  struct sockaddr_in sa;
+
+  // inet_pton returns 1 for success 0 for failure
+  if (inet_pton(AF_INET, ip_address, &(sa.sin_addr)) != 1) {
+    throw std::invalid_argument("Virtual ipv4 address is not in the expect format");
+  }
+}
+
+void ACA_ARP_Responder::_validate_ipv6_address(const char *ip_address)
+{
+  struct sockaddr_in sa;
+
+  // inet_pton returns 1 for success 0 for failure
+  if (inet_pton(AF_INET6, ip_address, &(sa.sin_addr)) != 1) {
+    throw std::invalid_argument("Virtual ipv6 address is not in the expect format");
+  }
+}
+
+int ACA_ARP_Responder::_validate_arp_entry(arp_config *arp_cfg_in)
+{
+  if (0 >= arp_cfg_in->mac_address.size()) {
+    throw std::invalid_argument("Input mac_string is null");
+  }
+
+  _validate_mac_address(arp_cfg_in->mac_address.c_str());
+
+  if (0 < arp_cfg_in->ipv4_address.size()) {
+    _validate_ipv4_address(arp_cfg_in->ipv4_address.c_str());
+  }
+
+  if (0 < arp_cfg_in->ipv6_address.size()) {
+    _validate_ipv6_address(arp_cfg_in->ipv4_address.c_str());
+  }
+
+  return EXIT_SUCCESS;
+}
+
+
+
+
+/************* Operation and procedure for dataplane *******************/
+
+
+void ACA_ARP_Responder::arp_recv(uint32_t in_port, void *vlan_hdr, void *message){
+  arp_message *arpmsg = nullptr;
+  vlan_message *vlanmsg = nullptr;
+
+  printf("receiving arp message\n");
+  if (!message) {
+    ACA_LOG_ERROR("%s", "ARP message is null!\n");
+    return;
+  }
+
+  vlanmsg = (vlan_message *)vlan_hdr; 
+  arpmsg = (arp_message *)message;
+
+
+  if(_validate_arp_message(arpmsg)){
+    ACA_LOG_ERROR("%s", "Invalid APR message!\n");
+    return;
+  }
+
+  _parse_arp_request(in_port, vlanmsg, arpmsg);
+
+  return;
+
+}
+void ACA_ARP_Responder::arp_xmit(uint32_t in_port, void  *vlanmsg, void *message){
+  arp_message *arpmsg = nullptr;
+  string bridge = "br-tun";
+  string inport = "in_port=controller";
+  string whitespace = " ";
+  string action = "actions=output:" + to_string(in_port);
+  string packetpre = "packet=";
+  string packet;
+  string options;
+
+  arpmsg = (arp_message *)message;
+  if(!arpmsg){
+    return;
+  }
+
+  packet = _serialize_arp_message((vlan_message *)vlanmsg, arpmsg);
+  if(packet.empty()){
```

Comment:
if this is an error condition, please log an error message and return error for this function